core/arch/arm/plat-k3/drivers/sa2ul.c

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Texas Instruments K3 SA2UL Driver
+ *
+ * Copyright (C) 2022 Texas Instruments Incorporated - https://www.ti.com/
+ *	Andrew Davis <afd@ti.com>
+ */
+
+#include <drivers/ti_sci.h>
+#include <initcall.h>
+#include <io.h>
+#include <keep.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
+#include <kernel/spinlock.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <rng_support.h>
+
+#include "sa2ul.h"
+
+#define	SA2UL_ES                0x0008
+#define SA2UL_ES_TRNG           BIT(3)
+#define	SA2UL_EEC               0x1000
+#define SA2UL_EEC_TRNG          BIT(3)
+
+#define FW_ENABLE_REGION        0x0a
+#define FW_BIG_ARM_PRIVID       0x01
+#define FW_WILDCARD_PRIVID      0xc3
+#define FW_SECURE_ONLY          GENMASK_32(8, 0)
+#define FW_NON_SECURE           GENMASK_32(16, 0)
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, SA2UL_BASE, SA2UL_REG_SIZE);
+
+static TEE_Result sa2ul_init(void)
+{
+	vaddr_t sa2ul = (vaddr_t)phys_to_virt(SA2UL_BASE, MEM_AREA_IO_SEC,
+					      RNG_REG_SIZE);
+	uint16_t fwl_id = SA2UL_TI_SCI_FW_ID;
+	uint16_t sa2ul_region = SA2UL_TI_SCI_FW_RGN_ID;
+	uint16_t rng_region = RNG_TI_SCI_FW_RGN_ID;
+	uint8_t owner_index = OPTEE_HOST_ID;
+	uint8_t owner_privid = 0;
+	uint16_t owner_permission_bits = 0;
+	uint32_t control = 0;
+	uint32_t permissions[FWL_MAX_PRIVID_SLOTS] = { };
+	uint64_t start_address = 0;
+	uint64_t end_address = 0;
+	uint32_t val = 0;
+	TEE_Result result = TEE_SUCCESS;
+	int ret = 0;
+
+	/* Power on the SA2UL device */
+	ret = ti_sci_device_get(SA2UL_TI_SCI_DEV_ID);
+	if (ret) {
+		EMSG("Failed to get SA2UL device");
+		return TEE_ERROR_GENERIC;
+	}
+
+	IMSG("Activated SA2UL device");
+
+	/* Try to claim the SA2UL firewall for ourselves */
+	ret = ti_sci_change_fwl_owner(fwl_id, sa2ul_region, owner_index,
+				      &owner_privid, &owner_permission_bits);
+	if (ret) {
+		/*
+		 * This is not fatal, it just means we are on an HS device
+		 * where the DMSC already owns the SA2UL. On GP we need
+		 * to do additional setup for access permissions below.
+		 */
+		DMSG("Could not change SA2UL firewall owner");
```

Comment:
On HS device, there will be some error printed from inside `ti_sci_change_fwl_owner`:

```
I/TC: Activated SA2UL device
E/TC:0 0 ti_sci_get_response:120 Message not acknowledged
E/TC:0 0 ti_sci_do_xfer:149 Failed to get response (-65536)
I/TC: Enabled firewalls for SA2UL TRNG device
```

This is really annoying. At the first glance, people would wonder if there is something wrong with their device.

Is it possible to check if the current device is not HS, then call this function?